File: internal/engine/baker/bands.go
package baker

// NOAA ENC navigational-purpose bands and the compilation-scale → band mapping.
// This lives in the (surviving, CGO-free) metadata package because grouping cells
// into bands for the tile57 per-band bake needs it independently of any tile
// baker. The slugs + zoom spans MUST match the frontend's CHART_BANDS so each
// per-band archive loads into its chart-<slug> source.

// Band is a NOAA ENC navigational-purpose band. Each bakes over its own zoom
// range and overzooms above max client-side.
type Band uint8

const (
	BandOverview Band = iota
	BandGeneral
	BandCoastal
	BandApproach
	BandHarbor
	BandBerthing
)

// BakeBand is one navigational-purpose band's identity for per-band archive
// baking: its frontend slug and native [Min,Max] zoom span.
type BakeBand struct {
	Slug     string
	Min, Max uint32
}

// BakeBands lists the bands coarse→fine for per-band archive baking — must match
// the frontend's CHART_BANDS (slug + zoom span) so each archive loads into its
// chart-<slug> source. The order matches the Band enum, so BakeBands()[Band] is
// that band's BakeBand.
func BakeBands() []BakeBand {
	return []BakeBand{
		{"overview", 0, 7},
		{"general", 7, 9},
		{"coastal", 9, 11},
		{"approach", 11, 13},
		{"harbor", 13, 16},
		{"berthing", 16, 18},
	}
}

// BandForScale maps a compilation-scale denominator (CSCL) to a band.
func BandForScale(cscl uint32) Band {
	n := cscl
	if n == 0 {
		n = 50_000
	}
	switch {
	case n <= 8_000:
		return BandBerthing
	case n <= 32_000:
		return BandHarbor
	case n <= 130_000:
		return BandApproach
	case n <= 500_000:
		return BandCoastal
	case n <= 2_300_000:
		return BandGeneral
	default:
		return BandOverview
	}
}

File: web/src/lib/bands.mjs
// bands.mjs — NOAA navigational-purpose "usage bands" (coarse→fine) shared across
// the web app: the band list, display colours/labels, the native Web-Mercator zoom
// range per band (mirrors CHART_BANDS in chart-canvas.mjs), and the scale↔band and
// zoom↔band mappings. One source of truth so the shell, the renderer overlays, and
// the feature controllers all agree.

export const BANDS = ["overview", "general", "coastal", "approach", "harbor", "berthing"];
export const BAND_LABEL = { overview: "Overview", general: "General", coastal: "Coastal", approach: "Approach", harbor: "Harbor", berthing: "Berthing" };
// Colour for the picker so overlapping cells are distinguishable and each band's
// toggle is identifiable.
export const BAND_COLOR = { overview: "#7e57c2", general: "#5c6bc0", coastal: "#26a69a", approach: "#9ccc65", harbor: "#ffa726", berthing: "#ef5350" };
// Native MIN display zoom per band. Below it a cell's chart detail isn't baked, so
// the shell draws its coverage outline instead. General is overzoomed out to z0 (it
// renders where no overview covers — see generalOverzoomMin in the baker).
export const BAND_MINZOOM = { overview: 0, general: 0, coastal: 9, approach: 11, harbor: 13, berthing: 16 };
// Native MAX display zoom per band. Drives the overscale cap: zooming past a band's
// native max + OVERSCALE_MARGIN over open water just enlarges blank water, so
// _updateZoomCap clamps to the finest band that actually covers the view.
export const BAND_MAXZOOM = { overview: 7, general: 9, coastal: 11, approach: 13, harbor: 16, berthing: 18 };
export const OVERSCALE_MARGIN = 1; // zoom-in levels allowed past the finest covering band (matches the bake default TILE57_FILLUP_DZ=1)
// Usage bands in coarse→fine order, for the dev band-filter rows.
export const DEV_BANDS = BANDS;

// A cell's compilation scale (1:N denominator) → its NOAA band. Mirrors
// bandForScale in the baker (bake.zig).
export function bandForScale(s) {
  const n = s || 0;
  if (n <= 8000) return "berthing";
  if (n <= 32000) return "harbor";
  if (n <= 130000) return "approach";
  if (n <= 500000) return "coastal";
  if (n <= 2300000) return "general";
  return "overview";
}

// A NOAA cell's usage-band digit (the 3rd char of e.g. "US5MD1MC") → its band, a
// fallback for the coverage overlay when the compilation scale isn't known yet (the
// NOAA catalogue hasn't loaded). The digit is a coverage-PURPOSE hint — bandForScale
// is authoritative — but good enough to colour/label a cell in the debug overlay.
export function bandForCellName(name) {
  const d = name && name.length >= 3 ? name.charCodeAt(2) - 48 : NaN;
  return BANDS[d - 1] || "harbor";
}

// The finest band whose source paints at zoom z (Band.zoomRange mins in bake.zig:
// overview 0, general 7, coastal 9, approach 11, harbor 13, berthing 16 — rounded
// to the client's display thresholds here).
export function bandForZoom(z) {
  if (z >= 16) return "berthing";
  if (z >= 14) return "harbor";
  if (z >= 12) return "approach";
  if (z >= 10) return "coastal";
  if (z >= 8) return "general";
  return "overview";
}
